### PR TITLE
remove discarded cf

### PR DESCRIPTION
**Pull request by zhangjinpeng1987**
List column families before open db, and drop discarded column families.

### CODE REVIEW

File: src/util/rocksdb.rs [hunk 1]
```diff
@@ -47,6 +48,52 @@ pub fn new_engine(path: &str, cfs: &[&str]) -> Result<DB, String> {
...
+
+    // Collect discarded column families.
+    let mut all_cfs_set = HashSet::new();
+    for cf in used_cfs {
+        all_cfs_set.insert(cf);
+    }
+    let mut discarded_cfs = vec![];
+    for cf in &cfs_list {
+        // Can't drop default column family
+        if !all_cfs_set.contains(&cf.as_str()) && cf != "default" {
```
Comment (siddontang) [inline src/util/rocksdb.rs]: use const var

File: src/util/rocksdb.rs [hunk 2]
```diff
@@ -47,6 +48,52 @@ pub fn new_engine(path: &str, cfs: &[&str]) -> Result<DB, String> {
...
 
+fn drop_discared_cfs(path: &str, used_cfs: &[&str]) -> Result<(), String> {
+    // List all column families in current db, if db not exist, it will return Err("IO error,
+    // XXXX/CURRENT: No such file or directory").
+    let mut opts = Options::new();
+    opts.create_if_missing(false);
+    let cfs_list = match DB::list_column_families(&opts, path) {
+        Ok(cfs) => cfs,
+        Err(e) => {
+            if e.starts_with("IO error") && e.ends_with("No such file or directory") {
```
Comment (siddontang) [inline src/util/rocksdb.rs]: magic error message...

**Comment by siddontang** [comment]
Do we need to support adding new cfs too?

File: src/util/rocksdb.rs [hunk 3]
```diff
@@ -47,6 +48,52 @@ pub fn new_engine(path: &str, cfs: &[&str]) -> Result<DB, String> {
...
+    };
+
+    // Collect discarded column families.
+    let mut all_cfs_set = HashSet::new();
+    for cf in used_cfs {
+        all_cfs_set.insert(cf);
+    }
+    let mut discarded_cfs = vec![];
+    for cf in &cfs_list {
+        // Can't drop default column family
```
Comment (siddontang) [inline src/util/rocksdb.rs]: we can check default at first.

File: src/util/rocksdb.rs [hunk 4]
```diff
@@ -47,6 +50,48 @@ pub fn new_engine(path: &str, cfs: &[&str]) -> Result<DB, String> {
...
 
+fn check_column_families(path: &str, used_cfs: &[&str]) -> Result<(), String> {
+    if !db_exist(path) {
+        return Ok(());
+    }
+
+    // List all column families in current db.
+    let mut opts = Options::new();
+    opts.create_if_missing(false);
+    let cfs_list = match DB::list_column_families(&opts, path) {
```
Comment (BusyJay) [inline src/util/rocksdb.rs]: Use `try!` instead.

File: src/util/rocksdb.rs [hunk 5]
```diff
@@ -47,6 +50,48 @@ pub fn new_engine(path: &str, cfs: &[&str]) -> Result<DB, String> {
...
+        Ok(cfs) => cfs,
+        Err(e) => return Err(e),
+    };
+
+    let existed: HashSet<&str> = cfs_list.iter().map(|v| v.as_str()).collect();
+    let needed: HashSet<&str> = used_cfs.iter().map(|v| *v).collect();
```
Comment (BusyJay) [inline src/util/rocksdb.rs]: s/used_cfs/using_cfs/ or s/used_cfs/needed_cfs/

File: src/util/rocksdb.rs [hunk 6]
```diff
@@ -47,6 +50,48 @@ pub fn new_engine(path: &str, cfs: &[&str]) -> Result<DB, String> {
...
+
+    // Open db.
+    let mut cfs_opts: Vec<Options> = vec![];
+    for _ in 0..cfs_list.len() {
+        cfs_opts.push(Options::new());
+    }
+    let cfs_ref_opts: Vec<&Options> = cfs_opts.iter().collect();
+    let cfs_list_str: Vec<&str> = cfs_list.iter().map(|cf| cf.as_str()).collect();
+    let mut db = DB::open_cf(&opts, path, &cfs_list_str, &cfs_ref_opts).unwrap();
```
Comment (BusyJay) [inline src/util/rocksdb.rs]: Can it cause data lose? For example, we may using different recovery mode than expected to open db.

File: src/util/rocksdb.rs [hunk 7]
```diff
@@ -47,6 +50,48 @@ pub fn new_engine(path: &str, cfs: &[&str]) -> Result<DB, String> {
...
+        // Never drop default column families.
+        if *cf != CF_DEFAULT {
+            try!(db.drop_cf(cf));
+        }
+    }
+
+    // Create needed column families not existed yet.
+    for cf in needed.difference(&existed) {
+        let opts = Options::new();
+        try!(db.create_cf(cf, &opts));
```
Comment (BusyJay) [inline src/util/rocksdb.rs]: Is it necessary to check whether cf is `CF_DEFAULT`?

File: src/util/rocksdb.rs [hunk 8]
```diff
@@ -80,3 +126,56 @@ pub fn new_engine_opt(mut opts: Options,
     }
     Ok(db)
 }
+
+fn db_exist(path: &str) -> bool {
+    let current_file = format!("{}/CURRENT", path);
```
Comment (BusyJay) [inline src/util/rocksdb.rs]: Why `CURRENT`?

File: src/util/rocksdb.rs (hunk 7 above)
Comment (zhangjinpeng1987) [inline src/util/rocksdb.rs]: If db has been created, the existed must contains CF_DEFAULT, so it is impossible cf == CF_DEFAULT.

**Comment by siddontang** [comment]
LGTM

File: src/util/rocksdb.rs [hunk 9]
```diff
@@ -55,9 +97,10 @@ pub fn new_engine_opt(mut opts: Options,
     // Currently we support 1) Create new db. 2) Open a db with CFs we want. 3) Open db with no
     // CF.
     // TODO: Support open db with incomplete CFs.
```
Comment (disksing) [inline src/util/rocksdb.rs]: Please update this comment.

File: src/util/rocksdb.rs [hunk 10]
```diff
@@ -80,3 +123,68 @@ pub fn new_engine_opt(mut opts: Options,
...
+fn db_exist(path: &str) -> bool {
+    let path = Path::new(path);
+    if !path.exists() || !path.is_dir() {
+        return false;
+    }
+
+    // If path is not an empty directory, we say db exists. If path is not an empty directory
+    // but db has not been created, DB::list_column_families will failed and we can cleanup
+    // the directory by this indication.
+    for _ in fs::read_dir(&path).unwrap() {
```
Comment (BusyJay) [inline src/util/rocksdb.rs]: fs.read_dir(&path).unwrap().next().is_some()

File: src/util/rocksdb.rs [hunk 11]
```diff
@@ -47,6 +51,44 @@ pub fn new_engine(path: &str, cfs: &[&str]) -> Result<DB, String> {
     new_engine_opt(opts, path, cfs, cfs_opts)
 }
 
+fn check_column_families(opts: &Options, path: &str, needed_cfs: &[&str]) -> Result<(), String> {
+    if !db_exist(path) {
```
Comment (BusyJay) [inline src/util/rocksdb.rs]: Why not create all the column family when db does not exist?

**Pull request merged**